Exec post adaptivity

nuclearkevin opened the pull request:
Closes #30547 

@GiudGiud

nuclearkevin committed:
Add the `POST_ADAPTIVITY` flag.
Ref. #30547

nuclearkevin committed:
Integrate `EXEC_POST_ADAPTIVITY` into multi-apps and multi-app transfers.
Ref. #30547

nuclearkevin committed:
Don't execute on initial adaptivity.
- MultiApps aren't constructed by then
- Ref. #30547

nuclearkevin committed:
Add a test for EXEC_POST_ADAPTIVITY.
Ref. #30547

GiudGiud reviewed (commented):
let's add some documentation to SetupInterface.md
notably the facts that:
- does not run on initial adaptivity
- does not run on every step of adaptivity, just after all steps have ran
- always runs even if adaptivity is skipped
- it is roughly located before TIMESTEP_END

nuclearkevin committed:
Add documentation in SetupInterface.md for `POST_ADAPTIVITY`.
Ref. #30547

Apply suggestions from code review.

Fix documentation failures

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Add additional documentation on the execution timing of `POST_ADAPTIVITY`.
Ref. #30547

GiudGiud commented on test/tests/multiapps/post_adapt_transfer/tests:
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#30547'
+  design = '/syntax/MultiApps/index.md'
```
```suggestion
  design = '/syntax/MultiApps/index.md SetupInterface.md'
```

GiudGiud commented on test/tests/multiapps/post_adapt_transfer/tests:
```diff
@@ -0,0 +1,11 @@
+    input = 'parent.i'
+    csvdiff = 'parent_out_sub0.csv'
+    requirement = 'The system shall allow for the execution of multi-apps and transfers after '
+                  'adaptivity and before the problem progresses to the next timestep.'
```
```suggestion
                  'adaptivity in the parent application and before the parent application progresses to the next timestep.'
```

GiudGiud reviewed (commented):
I ll try this locally. This looks good to go

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2882281/), step Docs: sync website on 3cd32fb wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30551/site)

This comment will be updated on new commits.

GiudGiud commented on test/tests/multiapps/post_adapt_transfer/tests:
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#30547'
+  design = '/syntax/MultiApps/index.md SetupInterface.md'
```
```suggestion
  design = 'syntax/MultiApps/index.md SetupInterface.md'
```

GiudGiud commented on framework/doc/content/source/interfaces/SetupInterface.md:
```diff
@@ -21,6 +21,7 @@ The default (and thus most common) flags are listed below:
 `NONLINEAR_CONVERGENCE` | Prior to each nonlinear convergence check. Note that `NONLINEAR` is not appropriate for this purpose since that only executes if it has already been determined that another nonlinear step is to be taken.
 `TIMESTEP_END` | After the solve for each time step.
 `TIMESTEP_BEGIN` | Prior to the solve for each time step.
+`POST_ADAPTIVITY` | Once all adaptivity steps have been performed on a timestep (after `TIMESTEP_END` and before progressing to the next step). Note that `POST_ADAPTIVITY` is not executed on initial adaptivity and will be executed even if the mesh is not adapted.
```
another peculiarity is this is skipped on the final time step (for example time step 3 when num_steps = 3) when adaptivity is skipped too

is that acceptable for your use case?

GiudGiud replied:
you could do 'EXEC_POST_ADAPTIVITY FINAL'

nuclearkevin replied:
Yep, using both flags will satisfy our requirements!

GiudGiud replied:
another note to add is that adaptivity is executed outside of multiapp fixed point iterations (when you set fixed_point_max_its in the Executioner). So you cant do fixed point iterations with this flag
Do you need fixed point iterations?

nuclearkevin replied:
We use pseudo-fixed point iteration in Cardinal (`TransientMultiApps` with `Transient` executioners and steady-state detection) as opposed to setting up fixed point iteration in an executioner the proper way, so this should be fine.

GiudGiud replied:
ok let's add both these caveats in the text please

loganharbour commented:
@nuclearkevin just added you to the repository for automatic testing

GiudGiud reviewed (approved):
squashed the fixups

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2887843/), step Generate coverage on 1f047b9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e67d42</th>
      <th colspan=4>#30551 1f047b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.48%</td>
      <td>85.48%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112711</td>
      <td>112716</td>
      <td>+5</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19147</td>
      <td>19147</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30551/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30551/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30551/coverage/xfem)

</details>



This comment will be updated on new commits.
